File: server/python/src/iam/schemas/role.py
"""
角色相关 Pydantic Schemas
"""

from datetime import datetime

from framework.schemas import BaseModel, BasePaginatedQuery, BaseQuery
from pydantic import Field
from iam.schemas.permission import PermissionResponse


class RoleCreate(BaseModel):
    """角色创建请求"""

    code: str = Field(..., min_length=1, max_length=50, description="角色编码")
    name: str = Field(..., min_length=1, max_length=100, description="角色名称")
    description: str | None = Field(None, description="角色描述")


class RoleUpdate(BaseModel):
    """角色更新请求"""

    name: str | None = Field(None, max_length=100, description="角色名称")
    description: str | None = Field(None, description="角色描述")


class RolePermissionRequest(BaseModel):
    """角色权限分配请求"""

    permission_ids: list[str] = Field(..., description="权限 ID 列表")


class RoleQuery(BaseQuery):
    """角色列表查询参数"""

    pass


class RolePaginatedQuery(RoleQuery, BasePaginatedQuery):
    """角色分页查询参数"""

    pass


class RoleResponse(BaseModel):
    """角色视图对象"""

    id: str
    tenant_id: str | None
    code: str
    name: str
    description: str | None
    is_system: bool
    created_at: datetime


class RolePaginatedListResponse(BaseModel):
    """角色分页列表响应"""

    total: int
    page: int
    page_size: int
    items: list[RoleResponse]


class RoleWithPermissionsResponse(BaseModel):
    """角色及其权限视图对象"""

    id: str
    tenant_id: str | None
    code: str
    name: str
    description: str | None
    is_system: bool
    permissions: list[PermissionResponse]
    created_at: datetime


class RoleOptionResponse(BaseModel):
    """角色选项视图对象（下拉列表用）"""

    id: str
    code: str
    name: str
    description: str | None


class RoleMemberAssignRequest(BaseModel):
    """角色成员分配请求"""

    user_ids: list[str] = Field(..., min_length=1, description="用户 ID 列表")


class RoleMenuAssignRequest(BaseModel):
    """角色菜单分配请求"""

    menu_ids: list[str] = Field(..., description="菜单 ID 列表")


class RoleMemberResponse(BaseModel):
    """角色成员视图对象"""

    user_id: str
    username: str
    nickname: str | None
    email: str | None
    phone: str | None
    status: str

File: web/vue/src/iam/api/role.ts
import { del, get, post, put } from "@/framework/api/client";
import type { Success, SuccessExtra } from "@/framework/types";
import type { Permission, Role, RolePaginatedQuery } from "@/iam/types";

export interface RoleCreate {
  code: string;
  name: string;
  description?: string;
}

export interface RoleUpdate {
  name?: string;
  description?: string;
}

export interface AssignPermissionsParams {
  permission_ids: string[];
}

/**
 * 获取角色列表
 */
export const getRoles = (params?: RolePaginatedQuery) =>
  get<SuccessExtra<Role[]>>("/iam/admin/v1/roles", { params });

/**
 * 获取角色详情（含权限）
 */
export const getRole = (id: string) =>
  get<Success<Role>>(`/iam/admin/v1/roles/${id}`);

/**
 * 创建角色
 */
export const createRole = (data: RoleCreate) =>
  post<Success<Role>>("/iam/admin/v1/roles", data);

/**
 * 更新角色
 */
export const updateRole = (id: string, data: RoleUpdate) =>
  put<Success<Role>>(`/iam/admin/v1/roles/${id}`, data);

/**
 * 删除角色
 */
export const deleteRole = (id: string) =>
  del<Success<void>>(`/iam/admin/v1/roles/${id}`);

/**
 * 为角色分配权限
 */
export const assignRolePermissions = (role_id: string, data: AssignPermissionsParams) =>
  post<Success<void>>(`/iam/admin/v1/roles/${role_id}/permissions`, data);

/**
 * 获取角色权限
 */
export const getRolePermissions = (role_id: string) =>
  get<Success<Permission[]>>(`/iam/admin/v1/roles/${role_id}/permissions`);

/**
 * 获取角色成员列表
 */
export const getRoleMembers = (role_id: string) =>
  get<Success<{ user_id: string; username: string; nickname?: string; email?: string; phone?: string; status: string }[]>>(`/iam/admin/v1/roles/${role_id}/members`);

/**
 * 为角色添加成员
 */
export const addRoleMembers = (role_id: string, user_ids: string[]) =>
  post<Success<{ added: number }>>(`/iam/admin/v1/roles/${role_id}/members`, { user_ids });

/**
 * 移除角色成员
 */
export const removeRoleMember = (role_id: string, user_id: string) =>
  del<Success<void>>(`/iam/admin/v1/roles/${role_id}/members/${user_id}`);

/**
 * 获取角色菜单
 */
export const getRoleMenus = (role_id: string) =>
  get<Success<string[]>>(`/iam/admin/v1/roles/${role_id}/menus`);

/**
 * 为角色分配菜单
 */
export const assignRoleMenus = (role_id: string, menu_ids: string[]) =>
  post<Success<void>>(`/iam/admin/v1/roles/${role_id}/menus`, { menu_ids });
